### PR TITLE
📊  war: add cow dataset

### PR DESCRIPTION
**Pull request by lucasrodes**
parent issue: https://github.com/owid/owid-issues/issues/446

### 📝 Notes
_(mostly for @bastianherre)_

- CoW dataset provides four tables. One for each 'conflict type': intra-state, inter-state, extra-state and non-state conflicts.
- Mixed usage of regions:
  - extra- and non-state tables use the same six regions: Americas, Asia, Africa, Europe, Oceania, and Middle East.
  - inter-state uses these regions but adds "composite regions" (e.g. "Europe and Asia"). Therefore, we map deaths in composite regions to the individual regions (evenly) and count conflicts in both regions.
  - intra-state regions use a similar set of regions as the other conflicts but also use several composite regions. For intra-state conflicts, all deaths should belong to the same region (as the conflict occurs within a state).
- Missing values for deaths: Each table has some missing values.
  - extra-, inter- and extra-state tables present some nulls in the deaths column.
  - non-state presents deaths in two columns (state, and non-state deaths). Sometimes one or the other is null.
  - Full details on these nulls in this [spreadsheet](https://docs.google.com/spreadsheets/d/1Z8zujPQoJ6usWXR4Rs6vBVefM0HQAmheeEHvF01Yqh0/edit#gid=1866531737).
  - For now, I've just set these nulls to zero.

### Code review
_(mostly for reviewer of the code)_

Some help notes for review.

- Adds garden (`data://garden/war/2023-07-18/cow`) and grapher (`data://grapher/war/2023-07-26/cow`) steps
- COW dataset is built from four different tables: "extra-state conflicts", "non-state conflicts", "intra-state conflicts" and "inter-state conflicts". These tables have slightly different structure, hence we initially process these tables separately with functions `make_table_*` (e.g. `make_table_inter`). Then we combine all of these with `combine_tables`.
- Each conflict has a region assigned (where the conflict happened). Unfortunately, each of the tables uses slightly different regions (i.e. some use "sub-saharan africa", others just "Africa"). Hence, some work is done to standardise these regions so all tables use the same regions.
  - In this sense, we have a file `etl/steps/data/garden/war/2023-07-18/cow.intrastate.region_mapping_custom.json`, which helps map conflicts to regions where some manual work was needed (for intra-state regions).

I tried adding docstring to each function.

### CODE REVIEW

File: etl/steps/data/garden/war/2023-07-18/cow.meta.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+    It has been constructed using the Correlates of War - War data dataset.
+
+
+tables:
+  cow:
+    variables:
+      number_deaths_ongoing_conflicts:
+        title: Number of soldier deaths in ongoing conflicts
```
[the remark this answers is not in the document]
Comment (spoonerf) [reply etl/steps/data/garden/war/2023-07-18/cow.meta.yml:14]: Ah it looks like the values for intra-state are what I was looking for. I would have thought inter-state would have the higher values? If it means conflicts between countries, but perhaps I have misunderstood.

File: etl/steps/data/garden/war/2023-07-18/cow.meta.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+dataset:
+  title: History of war (COW, 2020)
+  description: >-
+    This dataset contains information on armed conflicts, both internal and external, in the period of X and 2007.
```
Comment (spoonerf) [inline etl/steps/data/garden/war/2023-07-18/cow.meta.yml:4]: Should the 'X' be here?

**Review by spoonerf — APPROVED** [review]
Hello, looks good, I had a couple of queries about the data but could just be me not understanding terms.

File: etl/steps/data/garden/war/2023-07-18/cow.meta.yml (hunk 2 above)
Comment (lucasrodes) [reply etl/steps/data/garden/war/2023-07-18/cow.meta.yml:4]: totally, changing this!

File: etl/steps/data/garden/war/2023-07-18/cow.meta.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucasrodes) [reply etl/steps/data/garden/war/2023-07-18/cow.meta.yml]: Brutal! Thanks, Fiona. There was a line of code that was leading to a significant bug in inter-state death count.

**Pull request merged**